### ISSUE TITLE
Uncaught exception from FUSE operation create

### ISSUE THREAD

**Issue by JeffPsycle** [rianhunter/dbxfs#17]
I am having more issues from Fuse since the 1.0.35/1.0.36 updates. 

Python 3.6.5
dbxfs 1.0.36
fusepyng 1.0.7

fuse-sshfs-2.10-1.fc28.x86_64
fuse-libs-2.9.7-11.fc28.x86_64
fuse-2.9.7-11.fc28.x86_64
fuse-common-3.2.1-11.fc28.x86_64

Currently I cannot write anything out to the path that Dropbox is mounted to. For example:

```
sudo -u dbuser touch /home/dbuser/dropbox/test.txt
touch: setting times of '/home/dbuser/dropbox/test.txt': No such file or directory
```

I am getting the following in /var/log/messages:

```
ERROR:fuse:Uncaught exception from FUSE operation create, returning errno.EINVAL: ApiError('54e2b948b2762292b1bcbfbfde688f12', UploadError('other', None))#012Traceback (most recent call last):#012  File "/usr/local/lib/python3.6/site-packages/fusepyng.py", line 815, in _wrapper#012    return func(*args, **kwargs) or 0#012  File "/usr/local/lib/python3.6/site-packages/fusepyng.py", line 1082, in create#012    return self.operations('create', path, mode, fi)#012  File "/usr/local/lib/python3.6/site-packages/fusepyng.py", line 1331, in __call__#012    ret = getattr(self, op)(path, *args)#012  File "/usr/local/lib/python3.6/site-packages/userspacefs/fuse_adapter.py", line 128, in create#012    fi.fh = self._save_file(self._fs.open(self._conv_path(path), fi.flags))#012  File "/usr/local/lib/python3.6/site-packages/dbxfs/translate_ignored_files.py", line 91, in open#012    return self._sub.open(self._convert_path(path), *n, **kw)#012  File "/usr/local/lib/python3.6/site-packages/dbxfs/cachingfs.py", line 1485, in open#012    directory=directory, only_cache=True)#012  File "/usr/local/lib/python3.6/site-packages/dbxfs/cachingfs.py", line 1630, in stat#012    only_cache=only_cache)#012  File "/usr/local/lib/python3.6/site-packages/dbxfs/cachingfs.py", line 1557, in _stat_repeat#012    new_stat = self._fs.x_stat_create(path, create_mode, directory)#012  File "/usr/local/lib/python3.6/site-packages/dbxfs/dbxfs.py", line 679, in x_stat_create#012    return md_to_stat(self._stat_create(path, mode & ~os.O_TRUNC, directory))#012  File "/usr/local/lib/python3.6/site-packages/dbxfs/dbxfs.py", line 656, in _stat_create#012    md = self._clientv2.files_upload(b'', str(path))#012  File "/usr/local/lib/python3.6/site-packages/dropbox/base.py", line 2277, in files_upload#012    f,#012  File "/usr/local/lib/python3.6/site-packages/dropbox/dropbox.py", line 296, in request#012    user_message_locale)#012dropbox.exceptions.ApiError: ApiError('54e2b948b2762292b1bcbfbfde688f12', UploadError('other', None))
```

**Comment by rianhunter** [comment]
Hmm I'm not able to reproduce this problem. It looks like there is an error coming from Dropbox, so I don't think this has to do with a regression in 1.0.36. Unfortunately the Dropbox error you pasted doesn't have any information about why `UploadError('other', None)`. Some thoughts/suggestions:

* Have you tried making a via the Dropbox website? Does it fail?
* Have you recently changed you Dropbox password?
* If the answer to both of the above questions is "no," try re-linking dbxfs. Just delete the configuration file located at `dbxfs --print-default-config-file` and re-run it.

**Comment by JeffPsycle** [comment]
@rianhunter 

Thanks for the fast response on this. Yesterday while I was scrambling I did test that I could upload to the website and I went through a fresh key generation. I can also verify that the password has not changed. This is really quite strange seeing as I can read any files from Dropbox, just not write.

Let me explain a bit more of our use case and see if that can put some light on this. I am using this on AWS with a combination of provisioned instances, some of which are attached to an auto scaling group that can bring up (and tear down) machines at any time based on system load others are just standard instances. We are using a script with the access_token_command in the config.json file which is on a network share so any new instance that comes up can use the key. I have also tried issuing new keys so that every instance had it's own but this did not change the ability to write. We also are using an EFS (nfs4) mount for the cache folder so all of the instances can use the same cache files.

Things I have tried to fix the issue:

- removing the EFS mount and using a local path for the cache
- issuing a new access_token
- provisioning a fresh instance using local cache and new access_token
- checking for files with null or control characters in the file names
- changing to our test Dropbox account and then changing back to the production account

Something to note, when I change to our test account I can write just fine.

**Comment by rianhunter** [comment]
Jeff let me first say that I love how you're using `dbxfs`. Your usage of `access_token_command` and putting the cache folder on a network mount is fine, and shouldn't cause any problems (though now this makes me want to remove the cache-folder auto-creation to account for cases where the mount temporarily goes down).

Ordinarily I would suspect it was an invalid path name but now I strongly suspect the Dropbox API is limiting one of your accounts for some unknown reason. Try this test script:

    #!/usr/bin/env python3
    import getpass
    import dropbox
    
    access_token = getpass.getpass("enter access token: ")
    path = input("enter path (start with /): ")
    
    client = dropbox.Dropbox(access_token)
    
    client.files_upload(b'', path)

What are the results of using that script with your different accounts? Different access tokens? If your production account is still blocked, then you may need to take it up with the Dropbox API team. It's possible your account flagged some limit if you were doing excessive file operations.

**Comment by greg-db** [comment]
The `other` generally indicates that the API is returning an error type that your copy of the Dropbox library doesn't understand. This generally happens when you have an older version of the library installed. Try upgrading `dropbox` to the latest version and try again for a better error message. 

([`dbxfs` only requires `dropbox >=3.38`](https://github.com/rianhunter/dbxfs/blob/master/setup.py#L40), but [the latest release is currently v9.1.0](https://github.com/dropbox/dropbox-sdk-python/releases).)

**Comment by rianhunter** [comment]
@greg-db At the time of the error he was running version 9.1.0 of the Dropbox Python SDK

**Comment by JeffPsycle** [comment]
@rianhunter Thank you so much for the snippet to assist in getting to the bottom of this. I was able to write with multiple test accounts but all of the access tokens assigned to our production account failed.

We are working with Dropbox support, which they are waffling a bit on what the cause of the problem is. I however think you are dead on that we hit the API call limit. Looking at the Dropbox Business Advanced plan:

```Transfer data from your existing solutions with 25,000 included API calls per month to data transport partners. For additional data transport needs, please contact support.```

I am hoping that the clock on that starts on the first of the month and tomorrow we magically see everything working again.

What with this new information I would consider this not to be an issue with dbxfs, but if you would like I will continue to comment until we find a solution before closing the issue.

**Comment by JeffPsycle** [comment]
@greg-db I have just caught up with our guys that were working directly with you. Thanks for the information on the issues with the API calls. I appreciate the direct answers we were able to get from you. I have added the details from your correspondence for reference.

@rianhunter Here are the details from the other errors that were coming across on Dropbox's side:

```
I also checked on your account on our side, and from what I can see, you're running in to two errors for the upload endpoint:

a) 'path/conflict/file': This indicates that there's a file in the way, at the path specified in the upload call, so the upload could not be completed. You should remove the existing file, if desired, or specify a different write_mode/autorename strategy.

b) 'too_many_write_operations': The 'too_many_write_operations' error indicates "lock contention". That's a result of how Dropbox works on the backend. This is a technical inability to make a modification in the account or shared folder at the time of the API call. This error indicates that there was simultaneous activity in the account or shared folder preventing your app from making the state-modifying call (e.g., adding, editing, moving, or deleting files) it is attempting. The simultaneous activity could be coming from your app itself, or elsewhere, e.g., from the user's desktop client. It can come from the same user, or another member of a shared folder.

You can find more information about lock contention here:

https://www.dropbox.com/developers/reference/data-ingress-guide

In short, to avoid this error, you should avoid making multiple concurrent state modifications. E.g., don't issue multiple such requests at a time. That won't guarantee that you won't run in to this error though, as contention can still come from other sources, so you may also want to implement automatic retries in your app. If you are sending these calls serially and are still seeing this error though, please try adding a short delay (e.g., a few seconds) before sending the next one.```

**Comment by austingoudge** [comment]
@rianhunter Can I just add my personal thanks to you for this project, and for your active responsiveness to the issues we have reported.

@greg-db Thank you for your direct responses to our issues, it's a breath of fresh air compared to the standard 'boilerplate' support we have received from DropBox in the past.  If you have any ability to feed back, I have a couple of suggestions though:

1. The DB helpdesk staff need to know that this API limit exists and they should have the ability to check the current status during the live chat session.  This would have helped save a lot of time in what is (still) a critical loss of service issue for us.
2. We are astonished that there is no warning of hitting these limits, either in advance or when it happens.  I would consider an alerting system to notify your clients essential.

**Comment by greg-db** [comment]
Thanks for the feedback! I'll share this internally.

Also, your team admins should have actually received warning emails when the API limit was being neared. (Perhaps they went to spam?)

**Comment by greg-db** [comment]
@rianhunter We were able to reproduce the issue with this error type not getting translated properly, so we're looking into that.

**Comment by rianhunter** [comment]
Cool thanks everyone for getting to the bottom of this and providing more information. In the meantime the conversation around improving error reporting for this case is happening here https://github.com/dropbox/dropbox-sdk-python/issues/157
